### DISCUSSION TITLE
Details on segment command

### DISCUSSION THREAD

**Discussion by akvilonBrown** [ivadomed/ivadomed#1289] [Q&A]
Greetings!

I’m trying to use **ivadomed** with FILM layers for nuclei segmentation from brightfield microscopy images deriving from different human tissues. The reason behind it is to check the performance and effect of metadata, such as simple categorical information about cell lines,  and compare it with my custom segmentation pipeline. There are circa 3000 images 1080x1080 px with corresponding nuclei masks, already split into train/validation/test sets. I took the train set and reorganized it according to the BIDS specifications with microscopy. I’m not sure whether it’s the optimal way because I put each image in a separate folder to imitate participant_id. Anyway, I was able to launch and complete training. It took more than a day, I did not use transformations except normalization, and my tiling was 512x512 px.
The evaluation of the model didn’t go smoothly - I got metrics in the log, but empty folders **pred_masks** and **results_eval**. However, I’m not interested in debugging it right now. I have a separate test set, and I wonder how to segment it. I failed to find details about _segment_ command, and I’m afraid of disrupting data and the model. Can you explain here how to point the trained model to input images the right way? What meaning do the settings have in the _config_ file during _segment_ command?

**Comment by jcohenadad** [comment]
Hi @akvilonBrown, thank you for reaching out!

Have you tried the [`ivadomed-segment-image`](https://ivadomed.org/scripts.html#ivadomed-segment-image)? It is specifically designed to run inference on single input images.

If you'd like to run the inference on a specific test folder (ie: not an a single image, but on an entire directory), you can use `ivadomed --test` as exemplified in https://ivadomed.org/tutorials/one_class_segmentation_2d_unet.html#evaluate-model.

**Comment by akvilonBrown** [comment]
Hello,  @jcohenadad 
Thank you for the prompt reply.
In _docs_ there is no example of using the command _ivadomed_segment_image_, but I'll try to figure it out.
I see it can handle multiple images, but specifying 500 items in a command is inconvenient. Anyway, I'll try it.
As for _ivadomed --test_, I assumed it it was designated to the same dataset, on which training was performed.
According to the config file, it is split into subsets, and after training, I run _ivadomed --test_ with the same config file, so it processes the test subset this time.
But do you claim that we can specify _path-data_ with another dataset?
Should I set therefore train/test fraction as 0/1 in the config file?

**Reply by jcohenadad** [reply]
Indeed, the doc was missing the example. It is now fixed (see [here](https://ivadomed.org/scripts.html#ivadomed-segment-image)). 

About the handling of 500 items, a more convenient approach would be to SHELL script the call, example:
```bash
FILES=`ls folder/*.nii`
for file in $FILES; do
  ivadomed_segment_image --image $file --model <PATH_TO_YOUR_MODEL_DIR>
done
```

About `ivadomed --test`: I'm not sure if specifying another path-data than the one you used for training would work. Could you please give it a try?

**Reply by hermancollin** [reply]
> As for _ivadomed --test_, I assumed it it was designated to the same dataset, on which training was performed.

You are correct.

**Comment by akvilonBrown** [comment]
Hello, @jcohenadad !
My _test_ command fails with the following errors when the progress bar with inference completes:
on the setup without FILM: _"KeyError: "None of ['image_id'] are in the columns"_
on the setup with FILM: _"ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (453,) + inhomogeneous part."_

I also tried the trick with a separate test set, so I modified the _config_ file and specified another folder in _path_data_.
It seems to start working in general, but I got the same errors as with previous datasets.
I set the train/test fraction as 0.1/0.8; otherwise, it fails earlier, demanding that some data should be allocated for train and validation parts. The data is preprocessed - near each PNG file, a NIFTI file appears.
_ivadomed_segment_image_ wasn't present in my installation **ivadomed 2.9.8**.
I have added the script from GitHub.
It works with the ordinary model (without FILM)
But fails in model with FILM:
```
in get_onehotencoder
    idx[i][MetadataKW.INPUT_METADATA][0][context[ConfigKW.FILMED_UNET][ModelParamsKW.METADATA]] = options.get(OptionKW.METADATA)
```
I used my custom metadata _cell_line_.
But I can't find such metaparameter in the header of the source NIFTI file. 
How is the model with FILM supposed to recognize the metadata? Should I append some other info to the script?

I also suspect that my test errors might be caused by inconsistency in metadata.
In _json_ file nearby each image, I put the following: 
`{"PixelSize": [0.093, 0.093], "PixelSizeUnits": "um", "CellLine": "NIH3T3"}`
In participants.tsv I put:
```
participant_id	cell_line
sub-test0000	NIH3T3
...				...
```
In config file 
```
    "FiLMedUnet": {
        "applied": true,
        "metadata": "cell_line",
        "film_layers": [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
```
		
And training went smoothly.	
Do my suspicions have grounds, or do you have any other clues on resolving these errors?

**Reply by jcohenadad** [reply]
@hermancollin could you please look into it?

**Comment by hermancollin** [comment]
Hi @akvilonBrown. Just a few things I would like to clarify:
1. ~~Have you tried @jcohenadad's suggestion above (with the script using `ivadomed_segment_image`)?~~ Edit: Ah I missed that part, I see that you did.
2. Would it be possible for you to provide the full error message when you try to run `ivadomed_segment_image`? 
3. To be clear, if I understand correctly, all you want is to segment your separate test set (i.e. no need for test script), is that right? Unfortunately, the `--segment` command  would require the data to be formatted according to BIDS. **However**, there are workarounds that we could use but I just want to make sure I understand the problem here. Is this test data in BIDS format?
4. Finally, if you are unsure about the validity of your BIDS dataset, I would suggest using https://bids-standard.github.io/bids-validator/ to make sure there are not unexpected errors.

**Comment by akvilonBrown** [comment]
Hello, @hermancollin. 
Thank you for chipping in.
I put a lot of information before, so it may look confusing. I just wanted to say that I tried all means to make inferences from the trained models.
Let me reintroduce my plot.

I train two models on microscopy images - one is regular Unet, another is Unet with FILM - to compare their performance afterward on the dedicated test set.
Datasets were formatted according to BIDS specs, and I used that online validator as well.  The test set also complies with BIDS (and it also gets PNG files cast to NIFTI with some preprocessing)
[Here](https://drive.google.com/file/d/1X6nU5bzAO-jLyE4jsmyCrg34srWLje3z/view?usp=share_link) is a shortened example of my dataset, along with config files.

Both plain and FILMed models completed training without errors (although I was not that happy that they split my train set further into subsets, making _train_ subset even smaller)
But both these models fail when I apply --_test_ command.

Plain Unet: folders **pred_masks** and **results_eval** created but no files within.
The log shows that inference took place and metrics were given, but then errors occurred.
```
Evaluation: 0it [00:00, ?it/s]
Traceback (most recent call last):
  File "/gpfs/space/home/iarosl82/ivadomed_env/bin/ivadomed", line 8, in <module>
    sys.exit(run_main())
  File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/main.py", line 623, in run_main
    run_command(context=context,
  File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/main.py", line 537, in run_command
    df_results = imed_evaluation.evaluate(bids_df, path_output=path_output,
  File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/evaluation.py", line 114, in evaluate
    df_results = df_results.set_index('image_id')
  File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/pandas/util/_decorators.py", line 311, in wrapper
    return func(*args, **kwargs)
  File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/pandas/core/frame.py", line 5503, in set_index
    raise KeyError(f"None of {missing} are in the columns")
KeyError: "None of ['image_id'] are in the columns"
```

FILMed Unet:
```
 File "/gpfs/space/home/iarosl82/ivadomed_env/bin/ivadomed", line 8, in <module>
    sys.exit(run_main())
  File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/main.py", line 623, in run_main
    run_command(context=context,
  File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/main.py", line 527, in run_command
    pred_metrics = imed_testing.test(model_params=model_params,
  File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/testing.py", line 77, in test
    preds_npy, gt_npy = run_inference(test_loader, model, model_params, testing_params, str(path_3Dpred),
  File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/testing.py", line 319, in run_inference
    save_film_params(gammas_dict, betas_dict, metadata_values_lst, model_params[ModelParamsKW.DEPTH],
  File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/loader/film.py", line 288, in save_film_params
    gammas_dict = {i: np.array(gammas[i]) for i in range(1, 2 * depth + 3)}
  File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/loader/film.py", line 288, in <dictcomp>
    gammas_dict = {i: np.array(gammas[i]) for i in range(1, 2 * depth + 3)}
ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (453,) + inhomogeneous part.
```

_ivadomed-segment-image_ works but only with "normal" Unet.  So I could use it as a workaround to segment all my test images in the loop as @jcohenadad suggested to get masks and evaluate it with my own scripts.
_ivadomed-segment-image_ fails with FILMed Unet. 

```
Traceback (most recent call last):
  File "/gpfs/space/home/iarosl82/ivadomed_env/bin/ivadomed_segment_image", line 8, in <module>
    sys.exit(main())
  File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/scripts/segment_image.py", line 101, in main
    segment_image(args.image, args.model, args.suffix, options)
  File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/scripts/segment_image.py", line 79, in segment_image
    nii_lst, target_lst = imed_inference.segment_volume(path_model, fname_images, options=options)
  File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/inference.py", line 483, in segment_volume
    onehotencoder = get_onehotencoder(context, folder_model, options, ds)
  File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/inference.py", line 113, in get_onehotencoder
    idx[i][MetadataKW.INPUT_METADATA][0][context[ConfigKW.FILMED_UNET][ModelParamsKW.METADATA]] = options.get(OptionKW.METADATA)
KeyError: 0
```



So, in summary, it would be great to address all errors but I would focus on the possibility to make inferences with separate dataset.
The usage of _--segment_ command is not given in docs.
_segment-image_ works only with normal Unet.
I have doubts that I did all correctly, especially with FILMed model.

**Reply by hermancollin** [reply]
> Both plain and FILMed models completed training without errors (although I was not that happy that they split my train set further into subsets, making _train_ subset even smaller) 

This is controlled by the `split_dataset` part of the config file. The split is train-validation-test. In your case, since you have a separate dataset for testing, it would indeed make sense to remove the examples in the testing set (either by specifying a split file or by tweaking the train-test fraction directly in the config). However, the validation set is essential for supervised learning.

## for default Unet

> Plain Unet: folders **pred_masks** and **results_eval** created but no files within. The log shows that inference took place and metrics were given, but then errors occurred.
> 
> ```
> Evaluation: 0it [00:00, ?it/s]
> Traceback (most recent call last):
>   File "/gpfs/space/home/iarosl82/ivadomed_env/bin/ivadomed", line 8, in <module>
>     sys.exit(run_main())
>   File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/main.py", line 623, in run_main
>     run_command(context=context,
>   File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/main.py", line 537, in run_command
>     df_results = imed_evaluation.evaluate(bids_df, path_output=path_output,
>   File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/evaluation.py", line 114, in evaluate
>     df_results = df_results.set_index('image_id')
>   File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/pandas/util/_decorators.py", line 311, in wrapper
>     return func(*args, **kwargs)
>   File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/pandas/core/frame.py", line 5503, in set_index
>     raise KeyError(f"None of {missing} are in the columns")
> KeyError: "None of ['image_id'] are in the columns"
> ```
> 

I was able to reproduce this bug. I really don't know why, but the prediction is not saved in the right folder (i.e. should be saved in **log_sevenlines_train/pred_masks/** but are saved in the parent folder **log_sevenlines_train/** instead). A temporary fix is to move the `pred_class-0_pred.png` and `pred_pred.nii.gz` files inside **log_sevenlines_train/pred_masks/** and run the `--test` command again. I'm sorry for this inconvenience, I have never seen this behavior before.

With that being said, I was successfully able to segment the whole dataset with your config file by simply specifying `ivadomed -c config_microscopy_train.json --segment`. Maybe the doc wasn't clear about this but this is the exact purpose of the `--segment` command: to do inference on an entire BIDS dataset. The predictions can be found in **log_sevenlines_train/pred_masks/**.

---
## for FILMed Unet


> ```
>  File "/gpfs/space/home/iarosl82/ivadomed_env/bin/ivadomed", line 8, in <module>
>     sys.exit(run_main())
>   File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/main.py", line 623, in run_main
>     run_command(context=context,
>   File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/main.py", line 527, in run_command
>     pred_metrics = imed_testing.test(model_params=model_params,
>   File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/testing.py", line 77, in test
>     preds_npy, gt_npy = run_inference(test_loader, model, model_params, testing_params, str(path_3Dpred),
>   File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/testing.py", line 319, in run_inference
>     save_film_params(gammas_dict, betas_dict, metadata_values_lst, model_params[ModelParamsKW.DEPTH],
>   File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/loader/film.py", line 288, in save_film_params
>     gammas_dict = {i: np.array(gammas[i]) for i in range(1, 2 * depth + 3)}
>   File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/loader/film.py", line 288, in <dictcomp>
>     gammas_dict = {i: np.array(gammas[i]) for i in range(1, 2 * depth + 3)}
> ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (453,) + inhomogeneous part.
> ```
> 

I was also able to reproduce this bug. For some reason, after inference is ran, the film parameters are saved. I'm not sure why this is done honestly. I was able to make the `--test` command run successfully by applying the small fixed I previously described (moving the predictions in the appropriate folder) AND commenting lines 318, 319 and 320 of `ivadomed/testing.py`. If you don't know how to do this, I can give you further details on how to change the code locally.

> _ivadomed-segment-image_ works but only with "normal" Unet. So I could use it as a workaround to segment all my test images in the loop as @jcohenadad suggested to get masks and evaluate it with my own scripts. _ivadomed-segment-image_ fails with FILMed Unet.
> 
> ```
> Traceback (most recent call last):
>   File "/gpfs/space/home/iarosl82/ivadomed_env/bin/ivadomed_segment_image", line 8, in <module>
>     sys.exit(main())
>   File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/scripts/segment_image.py", line 101, in main
>     segment_image(args.image, args.model, args.suffix, options)
>   File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/scripts/segment_image.py", line 79, in segment_image
>     nii_lst, target_lst = imed_inference.segment_volume(path_model, fname_images, options=options)
>   File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/inference.py", line 483, in segment_volume
>     onehotencoder = get_onehotencoder(context, folder_model, options, ds)
>   File "/gpfs/space/home/iarosl82/ivadomed_env/lib/python3.8/site-packages/ivadomed/inference.py", line 113, in get_onehotencoder
>     idx[i][MetadataKW.INPUT_METADATA][0][context[ConfigKW.FILMED_UNET][ModelParamsKW.METADATA]] = options.get(OptionKW.METADATA)
> KeyError: 0
> ```

Again, I was able to reproduce this using the `--segment` command. However, this might be more problematic and related to #940. Let me get back to you on this one. In the meantime, if the `--test` command works for prediction, we could use it as a workaround.

**Comment by akvilonBrown** [comment]
Hi, @hermancollin
I applied your suggestions, and they work.
I was able to run _--test_ and _--segment_ on default Unet and --test only on **FILMed Unet**.
The figures in **evaluation_3Dmetrics.csv** make no sense (Dice is very low) -  the output in the console is more reasonable.
Anyway, I can collect predictions from the _--segment_ command and evaluate them on my own.
But only for the default Unet.
The issue with **FILMedUnet** persists, indeed.
The _--test_ command gives only final figures but doesn't save predictions.
I require predictions for interpretability and individual inspection.
Actually, I doubt whether **FILMedUnet** utilizes my custom metadata, but it's hard to verify.
I can check it training on my custom synthetic dataset, where the ordinary model would perform worse since similar source images can have different targets depending on the metadata.

**Reply by hermancollin** [reply]
@akvilonBrown Did you monitor the training of the FILM model on wandb? If so, did you notice a good performance?

I suspect that the problems you are encountering with the FILM model were caused by recent changes in the loader. I'll check if it's possible to downgrade ivadomed and get it to work with your example data and I'll get back to you.

**Reply by akvilonBrown** [reply]
Yes, @hermancollin, it was successful, and scores climbed up, but pretty much the same as with default Unet.  Also, I just finished new FILMed model training with curated synthetic data, which is supposed to output different masks depending on metadata. It can't reach acceptable validation performance as if metadata is not used.  So I believe FILMed model just switches to the default mode - I have the same behavior in my own implementation when I supply dummy metadata.  I can share this synthetic data if you like to debug **ivadomed** model.

**Reply by hermancollin** [reply]
@akvilonBrown could you explain briefly the task you are trying to achieve with your metadata? Because if the expected output is drastically different depending on the metadata, I'm not sure FiLM would be very efficient here, hence a similar performance with the default Unet.

**Reply by akvilonBrown** [reply]
@hermancollin,  I'm following [the work of A. Lemay](https://proceedings.mlr.press/v143/lemay21a.html),  from which I learned about **ivadomed**.
She drew different segmentation masks of tumors depending on the prior knowledge about tumor type (metadata).
Similarly, I can segment either nuclei or anomalies (debris) on microscopy images depending on the metadata.
This approach has some benefits compared to the case with separate models.
These cases are artificial because we set that kind of metadata as a task, instructing what segmentation the model should produce.
It works, and it makes a big difference compared to the default model because the ordinary model cannot discriminate tasks. 
I'm also trying to find cases where natural metadata can help.
Mostly the difference is marginal, indeed.  
But anyway, it's for comparison between my implementation and FILM.

**Comment by akvilonBrown** [comment]
Hello!
Is there a tutorial for FiLMedUnet?  Or a slice of some dataset which was successfully used for FiLMedUnet training.  I need the example to copy the structure.

**Reply by hermancollin** [reply]
Hi @akvilonBrown. 

Last week, I went down the rabbit hole of trying to make the FiLMedUNet work using ivadomed. I more or less found a config that should be usable, but it used an earlier version of ivadomed. I tried downgrading but wasn't able to reproduce the training because of outdated dependencies. In theory, it would be possible to downgrade dependencies but it would be pretty complex to setup.

FiLM was introduced in ivadomed specifically for the study you mentioned. This was more than 2 years ago and no one from our lab really experimented further with FiLMedUNets afterwards, so the feature was let a little bit behind on the dev side. If you want to fix this and open a pull request to make the feature work, we will be more than happy to review it and it would also be an opportunity for you to be more "hands-on" with your project, knowing exactly how the FiLM layers are implemented. Otherwise, unfortunately, no one from our research lab currently has time to invest in fixing this, and I would recommend another alternative like [monai](https://monai.io/) (they have a FiLMedUNet implementation and maybe also a tutorial for it).

**Reply by akvilonBrown** [reply]
Hi, @hermancollin. I totally understand the situation that you sincerely described. No complaints. I have something similar in terms of time budget. Intuitively I belive that FiLM works, but something is wrong with catering metadata to it.  I understand the principle how it works, and it is not far from my models. But I have to find some third party implementation to verify my experiments. I'm familiar with MONAI, but fail to find FiLMed models there. Can you point?
Also, another question. I would like to make another swap and play with data mentioned by Lemay. I couldn't find links to the dataset "Multiclass Spinal Cord Tumor Segmentation." Is it public or proprietary?

**Reply by hermancollin** [reply]
Hi @akvilonBrown. You are correct in saying the FiLM models are unfortunately not currently implemented in monai. I was thinking of Attention Unet, which is not the same architecture. From a quick glance, I actually couldn't find another implementation of FiLM. It is a shame that we were the maintainers of the only open-source implementation. Don't get your hopes up, but I'll bring it up to our meetings to see if there's anything we could do. In principle, the FiLM layers are not too complex.

Regarding the spinal cord tumor dataset, it is hosted on our private data servers. Given that it contains scans from more than 300 human subjects, it's very likely that it's under medical ethics protections. If you are interested in using the data, maybe the lab would be open to a collaboration but I doubt you would be able to access it without signing NDAs.

**Reply by akvilonBrown** [reply]
No worries, @hermancollin. These things, FILM and dataset,  would be just supplementary to my project. I'll find a workaround. However, when some news comes up about FILM, please let me know.

**Reply by naga-karthik** [reply]
Thanks for initiating this wonderful discussion @akvilonBrown! I found two other options for you to consider - (1) This [paper](https://arxiv.org/pdf/1907.01277.pdf) uses a conditioned UNet for separating various musical instrument sources. They basically propose a control mechanism supported by the FiLM layers. You can find the [code](https://github.com/gabolsgabs/cunet) here, and (2) for a similar application here's is another [implementation](https://github.com/Veleslavia/conditioned-u-net) of the conditioned UNet. I hope these you find these useful!

Note that it might be as a plug-and-play as what we have in ivadomed, but these provide some alternatives for you to test out your hypotheses. This is just another suggestion - I noticed there was no mention of the word "conditioning" in this discussion. If you expand your search terms to include "conditioning", maybe you might have more luck find relevant papers!

**Reply by akvilonBrown** [reply]
@naga-karthik , excellent references, thank you!
